src/middleware.go

```diff
@@ -0,0 +1,48 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+)
+
+func APIPanicHandlerMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if err := recover(); err != nil {
+				log.Println(err)
+
+				js, err := json.Marshal(map[string]interface{}{"ok": false, "error": fmt.Sprint(err)})
+				if err != nil {
+					log.Println(err)
+					http.Error(w, "{\"ok\":false,\"error\":\"unknown\"}", http.StatusInternalServerError)
+					return
+				}
+
+				w.Header().Set("Content-Type", "application/json")
+				http.Error(w, string(js), http.StatusInternalServerError)
+			}
+		}()
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+func UIPanicHandlerMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if err := recover(); err != nil {
+				log.Println(err)
+
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := tmpl.ExecuteTemplate(w, "500", fmt.Sprint(err)); err != nil {
+					log.Println(err)
+					_, _ = w.Write([]byte("It failed :("))
```

Comment:
I think there is no need to handle the error of `w.Write`, because what can we do if there is an error?